KAFKA-4633: Always using regex pattern subscription in StreamThread

guozhangwang opened the pull request:
1. In StreamThread, always use subscribe(Pattern, ..) function in order to avoid sending MetadataRequest with specific topic names and cause brokers to possibly auto-create subscribed topics; the pattern is generated as "topic-1|topic-2..|topic-n".

2. In ConsumerCoordinator, let the leader to refresh its metadata if the generated assignment contains some topics that is not contained in the subscribed topics; also in SubscriptionState, modified the verification for regex subscription to against the regex pattern instead of the matched topics since the returned assignment may contain some topics not yet created when joining the group but existed after the rebalance; also modified some unit tests in `KafkaConsumerTest` to accommodate the above changes. 

3. Minor cleanup: changed String[] to List<String> to avoid overloaded functions.

4. Minor cleanup: enforced strong typing in SinkNodeFactory and removed unnecessary unchecked tags.

5. Minor cleanup: augmented unit test error message and fixed a potential transient failure in KafkaStreamTest.

guozhangwang committed:
enforce always using regex pattern subscription

mjsax commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -183,9 +183,9 @@ public void assignFromSubscribed(Collection<TopicPartition> assignments) {
-        for (TopicPartition tp : assignments)
-            if (!this.subscription.contains(tp.topic()))
-                throw new IllegalArgumentException("Assigned partition " + tp + " for non-subscribed topic.");
+//        for (TopicPartition tp : assignments)
```
Why remove this check? We still know all topics we want to subscribe to.

guozhangwang replied:
I'm still pondering on it, the issue is that during rebalance streams leader can possibly assign internal topic partitions while the consumer’s not aware of the topic yet (since these topics are not created in Kafka), the above check will fail. This is actually not introduced in this topic as even before we could hit this issue but it was not exposed by tests.

I think in general this check is valid, and in that case maybe I have to get an extended consumer class as we discussed offline..

guozhangwang replied:
@hachikuji @mjsax Updated to use the `subscriptionPattern` if possible; this is discussed offline and please see if it looks "simple enough".

guozhangwang commented:
Also by doing this PR now I see more failures that is being worked on with https://github.com/apache/kafka/pull/2371. This is because now Streams rebalance gets shorter and the race condition is then more likely to happen.

guozhangwang commented:
retest this please.

guozhangwang commented:
retest this please

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -160,37 +161,32 @@ public void addStateStore(String stateStoreName) {
+        private final Deserializer<?> valDeserializer;
 
-        private SourceNodeFactory(String name, String[] topics, Pattern pattern, Deserializer keyDeserializer, Deserializer valDeserializer) {
+        private SourceNodeFactory(String name, String[] topics, Pattern pattern, Deserializer<?> keyDeserializer, Deserializer<?> valDeserializer) {
```
Any reason why you don't make the topics param a list?

guozhangwang replied:
Input parameters are directly from the `addSource` calls, whose param is `String[]`. I feel like deferring the `asList` to its constructor than letting the caller to do it. No strong opinions though.

guozhangwang commented:
retest this please

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K4633-regex-pattern

guozhangwang committed:
use pattern to check assignment if possible; update metadata snapshot

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -307,13 +307,26 @@ public long timeToNextPoll(long now) {
 
         Map<String, Assignment> assignment = assignor.assign(metadata.fetch(), subscriptions);
 
+        // check if the assignor has realized some new topics, if yes update the snapshot
```
@hachikuji This is just to show my current thoughts about how to work around the unnecessary rebalance without changing `Assignor` APIs, this is really not that elegant but just as a demo, and I'm happy to revert this back for future general solution; my take is that if we would allow `assign()` functions to update the `metadatasnapshot` then we probably would need to change the API.

guozhangwang committed:
resolve conflicts rebasing from apache trunk

mjsax reviewed (approved):
LGTM for streams side. Cannot say anything about consumer changes.

dguy reviewed (approved):
LGTM - though i don't have an opinion on the changes to ConsumerCoordinator

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -307,13 +307,26 @@ public long timeToNextPoll(long now) {
 
+        // check if the assignor has realized some new topics, if yes update the snapshot
+        Map<String, Integer> partitionsPerTopic = new HashMap<>();
+        for (Map.Entry<String, Assignment> entry : assignment.entrySet()) {
```
This seems a bit hacky. Why not just refresh metadata to get accurate counts?

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K4633-regex-pattern

guozhangwang committed:
github comments

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -307,13 +312,40 @@ public long timeToNextPoll(long now) {
+        // own metadata with the newly added topics so that it will not trigger a subsequent rebalance
+        // when these topics gets updated from metadata refresh.
+        Set<String> assignedTopics = new HashSet<>();
+        for (String topic : assignment.keySet()) {
```
This doesn't seem right. The keys are the member ids, not topics, right?

guozhangwang replied:
my bad

guozhangwang committed:
minor fix

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K4633-regex-pattern

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -307,13 +312,41 @@ public long timeToNextPoll(long now) {
+        if (!assignedTopics.containsAll(allSubscribedTopics)) {
+            Set<String> notAssignedTopics = new HashSet<>(allSubscribedTopics);
+            notAssignedTopics.removeAll(assignedTopics);
+            log.warn("The following subscribed topics are not assigned to any members in the group: " + notAssignedTopics);
```
nit: use string interpolation. Also probably good to add the groupId to these messages.

guozhangwang committed:
integration test

guozhangwang committed:
make sure join repartition test will fail without this fix

guozhangwang committed:
github comments

guozhangwang commented:
@hachikuji I have addressed your comments, also turned on auto creation in `EmbeddedKafkaCluster` which is used in all Streams integration test, and have also modified `KStreamRepartitionJoinTest` to have 2 partitions for input topics and verified that without this fix, this test will always fail.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -183,9 +183,23 @@ public void assignFromSubscribed(Collection<TopicPartition> assignments) {
-        for (TopicPartition tp : assignments)
-            if (!this.subscription.contains(tp.topic()))
-                throw new IllegalArgumentException("Assigned partition " + tp + " for non-subscribed topic.");
+        if (this.subscribedPattern != null) {
```
Can we add a unit test covering the cases where the assignment contains an unsubscribed partition which matches and doesn't match the regex?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -307,13 +312,41 @@ public long timeToNextPoll(long now) {
+            log.warn("The following subscribed topics are not assigned to any members in the group {} : {} ", groupId, notAssignedTopics);
+        }
+
+        if (!allSubscribedTopics.containsAll(assignedTopics)) {
```
It would also be good to have a unit test which covers this case. We should ensure that 1) the leader retains the group subscription of all members that joined the group, and 2) that followers correctly add new topics to their `Metadata`.

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K4633-regex-pattern

guozhangwang committed:
add unit tests

guozhangwang commented:
@hachikuji comments addressed and unit tests added.

guozhangwang commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -215,6 +215,12 @@ protected void onJoinComplete(int generation,
 
         // update partition assignment
         subscriptions.assignFromSubscribed(assignment.partitions());
+        this.joinedSubscription = subscriptions.subscription();
```
I'm not sure this works. The purpose of the `joinedSubscription` field is to remember the exact list of topics that were used when joining the group. If a metadata update arrives after the rebalance has begun, then we can notice the fact that the joined subscription does not match the current subscription and we can trigger another rebalance. With this change, we will no longer be able to detect this case, which means that consumption from a topic matching the subscribed regex will be delayed (perhaps indefinitely). 

It seems the behavior we want is to only add to `joinedSubscription` those topics which were added to the assignment by the leader.

guozhangwang replied:
@hachikuji Here is my understanding, and correct me if I'm wrong:

1. In `assignFromSubscribed`, we will only change the `subscription` if it is regex pattern and the collected topics as in `topicsToSubscribe` is a superset of `this.subscription`.

2. Besides this added line, `joinedSubscription` is only set to `this.subscription` before sending the join group request, and the `this.subscription` itself is only modified either inside `subscribe()` calls on inside the previously line if the assigned topics is a super set of the subscription itself. 

So given these two, I think this line will only update `joinedSubscription` with those topics that were added to the assignment by the leader If that does happen, right?

hachikuji replied:
@guozhangwang A metadata fetch could return in the middle of a rebalance (i.e. with the JoinGroup on the wire). In this case, we may end up discovering a new topic that we didn't know about before and adding it to our subscription (see `subscribeFromPattern`). We are able to catch this case currently because although the subscription has changed, `joinedSubscription` has not. With this patch, this is no longer true.

guozhangwang replied:
That makes sense. I can go ahead and tackle on this. 

Just clarifying: After the group has formed, both leader and follower can still trigger a rebalance: leader will trigger a rebalance if the existing topics number of partitions has changed (including the topic is deleted); follower will trigger a rebalance if the subscription has changed (both due to a metadata refresh with regex pattern or user called `subscribe` again). Is that right?

And if we change the consumer coordinator to allow passing regex to the leader, I think `joinedSubscription` can be removed completely and only leader need to trigger rebalances unless users call `subscribe` again on any of the consumer member, is that right?

hachikuji replied:
> Just clarifying: After the group has formed, both leader and follower can still trigger a rebalance: leader will trigger a rebalance if the existing topics number of partitions has changed (including the topic is deleted); follower will trigger a rebalance if the subscription has changed (both due to a metadata refresh with regex pattern or user called subscribe again). Is that right?

Yes, right.

> And if we change the consumer coordinator to allow passing regex to the leader, I think joinedSubscription can be removed completely and only leader need to trigger rebalances unless users call subscribe again on any of the consumer member, is that right?

The leader will still have to deal with the potential for a metadata update during a rebalance, so I'm not sure we can remove `joinedSubscription`. At least we won't need this funky logic to try to change `joinedSubscription` after the rebalance though.

guozhangwang committed:
address Jason's comments

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -215,7 +215,27 @@ protected void onJoinComplete(int generation,
+        // TODO this part of the logic should be removed once we allow regex on leader assign and remove joinedSubscription
+        Set<String> addedTopics = new HashSet<>();
+        for (TopicPartition tp : subscriptions.assignedPartitions()) {
+            if (!subscriptions.subscription().contains(tp.topic()))
```
I'm puzzling a bit over whether we should be using `joinedSubscription` here instead of `subscriptions.subscription()`, or whether it matters. 

Seems it should be `joinedSubscription`. Suppose that `joinedSubscription` contained only [A] at the time of rebalance. Topic B is then created by the leader and assigned. Before the rebalance completes, the consumer discovers topic B, and `subscription` is updated to [A, B], while `joinedSubscription` is still [A]. The consumer then receives the assignment containing partitions from both A and B, but `addedTopics` will be empty (since B was already added to `subscription`). The consumer will then notice that `joinedSubscription` is [A], while `subscription` is [A, B], and request an unneeded rebalance.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -215,7 +215,27 @@ protected void onJoinComplete(int generation,
+
+        if (!addedTopics.isEmpty()) {
+            Set<String> newSubscription = new HashSet<>(subscriptions.subscription());
+            Set<String> newJoinedSubsciprtion = new HashSet<>(joinedSubscription);
```
nit: typo `newJoinedSubsciprtion`

guozhangwang committed:
use joinedSubscription

guozhangwang commented:
@hachikuji Found an issue with `ConsumerCoordinatorTest` while working on it, that subscription is updated on the leader and hence `addedTopics` will be empty; thus we need to use `joinedSubscription` to determine the added topics and use that to add to both subscriptions.

Could you please take another look?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -216,6 +216,32 @@ protected void onJoinComplete(int generation,
 
+        // check if the assignment contains some topics that is not in the original
+        // subscription, if yes we will obey what leader has decided and add these topics
+        // into the subscriptions
```
Add to the end, "as long as they still match the subscribed pattern"

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -216,6 +216,32 @@ protected void onJoinComplete(int generation,
         // update partition assignment
         subscriptions.assignFromSubscribed(assignment.partitions());
 
+        // check if the assignment contains some topics that is not in the original
```
nit: "is" -> "were"

guozhangwang committed:
github comments

guozhangwang commented:
@hachikuji Addressed, thanks.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -300,18 +307,18 @@ public void testJoinGroupInvalidGroupId() {
         final String consumerId = "leader";
 
-        subscriptions.subscribe(singleton(topicName), rebalanceListener);
+        subscriptions.subscribe(Pattern.compile("test.*"), rebalanceListener);
```
Not sure why this was changed. Do we have "normal" subscription covered elsewhere?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -300,18 +307,18 @@ public void testJoinGroupInvalidGroupId() {
-        Map<String, List<String>> memberSubscriptions = Collections.singletonMap(consumerId, singletonList(topicName));
-        partitionAssignor.prepare(Collections.singletonMap(consumerId, singletonList(tp)));
+        Map<String, List<String>> memberSubscriptions = Collections.singletonMap(consumerId, singletonList(topic1));
+        partitionAssignor.prepare(Collections.singletonMap(consumerId, Arrays.asList(t1p, t2p)));
```
Related to comment above. Can we do this in a separate test case? This is not the "normal" case, but it should be covered somewhere.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -307,13 +333,41 @@ public long timeToNextPoll(long now) {
 
         Map<String, Assignment> assignment = assignor.assign(metadata.fetch(), subscriptions);
 
+        // user-customized assignor may have created some topics that are not in the subscription list
```
Can we mention in this comment that this is a hack and not something we want to support long-term unless we push regex into the protocol?

hachikuji reviewed (commented):
Minor comments. This is looking as good as it will, I guess.

guozhangwang committed:
github comments

guozhangwang commented:
@hachikuji addressed.

hachikuji reviewed (approved):
LGTM
